### PR TITLE
Implant web proxy support

### PR DESCRIPTION
**Pull request by ItzMiracleOwO**
Migrate from #1554 
(Sorry for multiple PRs, I had to resolve the conflicts)

> Modified `OkHttpClient` to implent a web proxy method that is already wanted since #409 
> 
> I only modified configs for `zh` and `en`, because I do not recognize other languages.

### CODE REVIEW

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 1]
```diff
@@ -140,7 +145,7 @@
  * @see #sendJoinMessage(Player, String)
  * @see #sendLeaveMessage(Player, String)
  */
-@SuppressWarnings({"unused", "WeakerAccess", "ConstantConditions"})
+@SuppressWarnings({ "unused", "WeakerAccess", "ConstantConditions" })
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:148]: Please do not make all these unrelated style changes

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 2]
```diff
@@ -712,43 +834,97 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
...
-                )
-                .build();
+        String proxyHost = config.getString("ProxyHost").trim();
+        String proxyPort = config.getString("ProxyPort").trim();
+        final String authUser = config.getString("ProxyUser").trim();
+        final String authPassword = config.getString("ProxyPw").trim();
+
+        OkHttpClient httpClient; // initialized below
+        if (proxyHost != null && !proxyHost.isEmpty() && authPassword == null && authPassword.isEmpty()) {
+            System.setProperty("jdk.http.auth.tunneling.disabledSchemes", "");
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:854]: This property shouldn't be changed, it can be changed to be blank if needed by the user

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 3]
```diff
@@ -712,43 +834,97 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
...
+                    .dns(dns)
+                    // more lenient timeouts (normally 10 seconds for these 3)
+                    .connectTimeout(20, TimeUnit.SECONDS)
+                    .readTimeout(20, TimeUnit.SECONDS)
+                    .writeTimeout(20, TimeUnit.SECONDS)
+                    .hostnameVerifier(noopHostnameVerifier.isPresent() && noopHostnameVerifier.get()
+                            ? (hostname, sslSession) -> true
+                            : OkHostnameVerifier.INSTANCE)
+                    .build();
+        }
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:907]: The common code should not be duplicated, sample of how it should be structured
```java
OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder()
    .dispatcher(dispatcher)
    ...

if (proxyIsInUse) {
    builder = builder.proxy(...);
    if (proxyHasPassword) {
        builder = builder.proxyAuthenticator(...);
    }
}

httpClient = httpClientBuilder.build();
```

File: src/main/resources/config/en.yml [hunk 4]
```diff
@@ -8,6 +8,14 @@ ConfigVersion: ${version}
...
 BotToken: "BOTTOKEN"
 
+# Proxy settings, leave empty if you dont need this
+# You must restart your server after changing this option
+ProxyHost: ""
+ProxyPort: ""
+ProxyUser: ""
+ProxyPw: ""
+
+
```
Comment (Vankka) [inline src/main/resources/config/en.yml:18]: No double newline here,
```suggestion

```

File: src/main/resources/config/en.yml [hunk 5]
```diff
@@ -8,6 +8,14 @@ ConfigVersion: ${version}
 # You must restart your server after changing this option
 BotToken: "BOTTOKEN"
 
+# Proxy settings, leave empty if you dont need this
+# You must restart your server after changing this option
+ProxyHost: ""
+ProxyPort: ""
+ProxyUser: ""
+ProxyPw: ""
```
Comment (Vankka) [inline src/main/resources/config/en.yml:16]: This option should be `ProxyPassword` instead

File: src/main/resources/config/en.yml [hunk 6]
```diff
@@ -8,6 +8,14 @@ ConfigVersion: ${version}
 # You must restart your server after changing this option
 BotToken: "BOTTOKEN"
 
+# Proxy settings, leave empty if you dont need this
+# You must restart your server after changing this option
+ProxyHost: ""
+ProxyPort: ""
```
Comment (Vankka) [inline src/main/resources/config/en.yml:14]: This option should be a number instead of a string

**Review by Vankka — CHANGES_REQUESTED** [review]
Thank you for your PR and sorry about the late response, there are some problems listed below that need to be addressed before this can be merged

**Comment by ItzMiracleOwO** [comment]
Thanks for the reply, it has been a while before the PR has been created, please wait for me to resolve the conflicts and fix the issues.

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java (hunk 2 above)
Comment (ItzMiracleOwO) [reply src/main/java/github/scarsz/discordsrv/DiscordSRV.java:854]: This is modified if proxy is enabled, see: https://stackoverflow.com/questions/41806422/java-web-start-unable-to-tunnel-through-proxy-since-java-8-update-111

File: src/main/resources/config/en.yml (hunk 6 above)
Comment (ItzMiracleOwO) [reply src/main/resources/config/en.yml:14]: Proxy host should could be set to a domain. I believe you mean the port setting should be a number?

**Comment by ItzMiracleOwO** [comment]
@Vankka Hi. Please check the new codes

File: src/main/resources/config/zh.yml [hunk 7]
```diff
@@ -8,7 +8,12 @@ ConfigVersion: ${version}
 # 更改此選項後必須重新啟動服務器
 BotToken: "BOTTOKEN"
 
-# 从游戏到不和谐的频道链接
+ProxyHost: "代理服务器地址"
+ProxyPort: 1234 # 代理服务器端口
+ProxyUser: "代理服务器用户名"
+ProxyPw: "代理服务器密码"
```
Comment (granny) [inline src/main/resources/config/zh.yml:14]:
```suggestion
ProxyHost: "https://example.com"
ProxyPort: 1234
ProxyUser: "USERNAME"
ProxyPassword: "PASSWORD"
```

File: src/main/resources/config/en.yml [hunk 8]
```diff
@@ -8,6 +8,11 @@ ConfigVersion: ${version}
 # You must restart your server after changing this option
 BotToken: "BOTTOKEN"
 
+ProxyHost: "Proxy Host"
+ProxyPort: 1234 # Proxy Port
+ProxyUser: "Your username"
+ProxyPassword: "SomePassword"
```
Comment (granny) [inline src/main/resources/config/en.yml:14]:
```suggestion
ProxyHost: "https://example.com"
ProxyPort: 1234
ProxyUser: "USERNAME"
ProxyPassword: "PASSWORD"
```

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 9]
```diff
@@ -721,19 +726,42 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
...
+        String proxyPort = config.getString("ProxyPort").trim();
+        final String authUser = config.getString("ProxyUser").trim();
+        final String authPassword = config.getString("ProxyPassword").trim();
+
+        OkHttpClient httpClient; // Initialized below
+        OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder()
                 .dispatcher(dispatcher)
                 .connectionPool(connectionPool)
                 .dns(dns)
-                // more lenient timeouts (normally 10 seconds for these 3)
```
Comment (granny) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:728]: add this comment back

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 10]
```diff
@@ -721,19 +726,42 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
         dispatcher.setMaxRequestsPerHost(20); // most requests are to discord.com
         ConnectionPool connectionPool = new ConnectionPool(5, 10, TimeUnit.SECONDS);
 
-        OkHttpClient httpClient = new OkHttpClient.Builder()
+        String proxyHost = config.getString("ProxyHost").trim();
+        String proxyPort = config.getString("ProxyPort").trim();
+        final String authUser = config.getString("ProxyUser").trim();
+        final String authPassword = config.getString("ProxyPassword").trim();
+
+        OkHttpClient httpClient; // Initialized below
```
Comment (granny) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:734]:
```suggestion
```
assign the variable on the same line where you initialize it.

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 11]
```diff
@@ -721,19 +726,42 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
...
+                    @Override
+                    public Request authenticate(Route route, Response response) throws IOException {
+                        String credential = Credentials.basic(authUser, authPassword);
+                        return response.request().newBuilder().header("Proxy-Authorization", credential).build();
+                    }
+                });
+            }
+        } 
+
+        httpClient = httpClientBuilder.build();
```
Comment (granny) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:764]:
```suggestion
        OkHttpClient httpClient = httpClientBuilder.build();
```
assign the variable on the same line where you initialize it.

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 12]
```diff
@@ -721,19 +726,42 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
         dispatcher.setMaxRequestsPerHost(20); // most requests are to discord.com
         ConnectionPool connectionPool = new ConnectionPool(5, 10, TimeUnit.SECONDS);
 
-        OkHttpClient httpClient = new OkHttpClient.Builder()
+        String proxyHost = config.getString("ProxyHost").trim();
+        String proxyPort = config.getString("ProxyPort").trim();
+        final String authUser = config.getString("ProxyUser").trim();
+        final String authPassword = config.getString("ProxyPassword").trim();
```
Comment (granny) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:732]:
```suggestion
        String proxyHost = config.getString("ProxyHost").trim();
        String proxyPort = config.getString("ProxyPort").trim();
        String authUser = config.getString("ProxyUser").trim();
        String authPassword = config.getString("ProxyPassword").trim();
```
you don't need `final`s here

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 13]
```diff
@@ -721,19 +726,42 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
...
                 .writeTimeout(20, TimeUnit.SECONDS)
                 .hostnameVerifier(noopHostnameVerifier.isPresent() && noopHostnameVerifier.get()
                         ? (hostname, sslSession) -> true
-                        : OkHostnameVerifier.INSTANCE
-                )
-                .build();
+                        : OkHostnameVerifier.INSTANCE);
+
+        // Reduce proxy username and password logic duplication
+        if (proxyHost != null && !proxyHost.isEmpty()) {
```
Comment (granny) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:747]: `proxyHost` is always a `String`. Add a check against the default value of `https://example.com` instead.

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 14]
```diff
@@ -721,19 +726,42 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
         dispatcher.setMaxRequestsPerHost(20); // most requests are to discord.com
         ConnectionPool connectionPool = new ConnectionPool(5, 10, TimeUnit.SECONDS);
 
-        OkHttpClient httpClient = new OkHttpClient.Builder()
+        String proxyHost = config.getString("ProxyHost").trim();
+        String proxyPort = config.getString("ProxyPort").trim();
```
Comment (granny) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:730]: use `getInt` for the port.

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 15]
```diff
@@ -721,19 +726,42 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
...
+                        : OkHostnameVerifier.INSTANCE);
+
+        // Reduce proxy username and password logic duplication
+        if (proxyHost != null && !proxyHost.isEmpty()) {
+            // This had to be set to empty string to avoid issue with basic auth
+            System.setProperty("jdk.http.auth.tunneling.disabledSchemes", "");
+            Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(proxyHost, Integer.parseInt(proxyPort)));
+            httpClientBuilder = httpClientBuilder.proxy(proxy);
+
+            if (authPassword != null && !authPassword.isEmpty()) {
```
Comment (granny) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:753]: `authPassword` is always a `String`. Only check if it's empty.

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 16]
```diff
@@ -721,19 +726,42 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
...
                 .hostnameVerifier(noopHostnameVerifier.isPresent() && noopHostnameVerifier.get()
                         ? (hostname, sslSession) -> true
-                        : OkHostnameVerifier.INSTANCE
-                )
-                .build();
+                        : OkHostnameVerifier.INSTANCE);
+
+        // Reduce proxy username and password logic duplication
+        if (proxyHost != null && !proxyHost.isEmpty()) {
+            // This had to be set to empty string to avoid issue with basic auth
```
Comment (granny) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:748]: Add the stackoverflow link in your comment

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 17]
```diff
@@ -721,19 +726,42 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
...
+            System.setProperty("jdk.http.auth.tunneling.disabledSchemes", "");
+            Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(proxyHost, Integer.parseInt(proxyPort)));
```
Comment (granny) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:750]: You don't need to convert it to an Int if you retrieve the config option as shown in the earlier comment.

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 18]
```diff
@@ -721,19 +726,42 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
...
                 .readTimeout(20, TimeUnit.SECONDS)
                 .writeTimeout(20, TimeUnit.SECONDS)
                 .hostnameVerifier(noopHostnameVerifier.isPresent() && noopHostnameVerifier.get()
                         ? (hostname, sslSession) -> true
-                        : OkHostnameVerifier.INSTANCE
-                )
-                .build();
+                        : OkHostnameVerifier.INSTANCE);
+
+        // Reduce proxy username and password logic duplication
```
Comment (granny) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:746]: remove this comment

**Review by granny — CHANGES_REQUESTED** [review]
I probably missed something since the comments are squished together.

You still need to include the config options in all the other config files. Don't worry about translating them, you can just copy the english translation across them all.

**Comment by ItzMiracleOwO** [comment]
@granny Issue resolved, please check

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 19]
```diff
@@ -721,7 +726,12 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
         dispatcher.setMaxRequestsPerHost(20); // most requests are to discord.com
         ConnectionPool connectionPool = new ConnectionPool(5, 10, TimeUnit.SECONDS);
 
-        OkHttpClient httpClient = new OkHttpClient.Builder()
+        String proxyHost = config.getString("ProxyHost").trim();
+        String proxyPort = config.getInt("ProxyPort").trim();
```
Comment (granny) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:730]:
```suggestion
        int proxyPort = config.getInt("ProxyPort");
```
I'm honestly surprised `trim()` worked here on `getInt()`.

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 20]
```diff
@@ -731,9 +741,27 @@ private List<InetAddress> lookupPublic(String host) throws UnknownHostException
                 .writeTimeout(20, TimeUnit.SECONDS)
                 .hostnameVerifier(noopHostnameVerifier.isPresent() && noopHostnameVerifier.get()
                         ? (hostname, sslSession) -> true
-                        : OkHostnameVerifier.INSTANCE
-                )
-                .build();
+                        : OkHostnameVerifier.INSTANCE);
+
+        if (proxyHost != "https://example.com") {
```
Comment (granny) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java:746]:
```suggestion
        if (!proxyHost.isEmpty() && !proxyHost.equals("https://example.com")) {
```
You need to use the `equals()` method attached to a string. It would probably be a good idea to check if it's empty as well, just in case someone sets it to `""` in an attempt to disable the feature.

File: src/main/resources/config/en.yml [hunk 21]
```diff
@@ -8,6 +8,11 @@ ConfigVersion: ${version}
 # You must restart your server after changing this option
 BotToken: "BOTTOKEN"
 
+ProxyHost: "https://example.com"
```
Comment (granny) [inline src/main/resources/config/en.yml:11]:
```suggestion
# The proxy to be used for Discord
# Leave this alone if you don't understand what it does
ProxyHost: "https://example.com"  
```
Needs a comment explaining to people what this is used for. Copy this english translation across all the config translations you can't translate.

**Review by granny — CHANGES_REQUESTED** [review]
Just a few more things. The english translation of the comment can be changed if you think of a better way to word it.

It would probably be a good idea to add debug lines stating something along the lines of "using proxy for Discord".

**Comment by ItzMiracleOwO** [comment]
@granny work done, also I attempted to fix the failed build due to error

image

**Comment by ItzMiracleOwO** [comment]
Synced develope branch changes to patch-1

**Comment by granny** [comment]
Thank you for your PR!
